### PR TITLE
Fix label docs (and a couple of misc bits)

### PR DESCRIPTION
**Pull request by vext01**
Main work is in https://github.com/CensoredUsername/dynasm-rs/commit/8d5f2f2b7b89e836e93c7a4653e1b97046281112, as discussed in #91.

I was unable to check the rendered output of my non-sourcefile doc changes (e.g. language ref), as I don't know how to generate the html :)

### CODE REVIEW

File: doc/langref_common.md [hunk 1]
```diff
@@ -106,28 +106,34 @@ dynasm!(ops
...
-Dynamic | `=>expr`     | `=>expr`
-Extern  | `-`          | `extern expr`
+Type    |  Kind   | Definition   | Reference
+--------|---------|--------------|-----------
+Local   | static  | `label:`     | `>label` or `<label`
+GLobal  | static  | `->label:`   | `->label`
+Dynamic | dynamic | `=>expr`     | `=>expr`
+Extern  | dynamic | `-`          | `extern expr`
+
+Static label kinds  ("Local" and "Global") have their addresses resolved at the
```
Comment (vext01) [inline doc/langref_common.md:120]: Is this true?

Comment (CensoredUsername) [reply doc/langref_common.md:120]: This doesn't seem correct. All label types are resolved at `commit` time. Extern is also not a dynamic label, it's something completely different (it is literally just targetting a given address).

The difference between static and dynamic labels is that the identifier (name) is fixed at rustc compile time for static labels, while you can create new identifiers (basically just an incrementing integer) at runtime for dynamic labels.

Comment (vext01) [reply doc/langref_common.md:120]: I must admit, there was a fair amount of guesswork going on here. I don't actually know how dynasm works under the hood!

What I'd like the docs to be able to tell me is which kinds of label can be looked up with `LabelRegistry::resolve_static()` and what kinds can be looked up with `LabelRegistry::resolve_dynamic()`.

I assume that `local` and `global` are for `resolve_static()`, and that `dynamic` is (guess what?) resolve_dynamic(). But what about `extern`?

**Comment by vext01** [comment]
I've pushed a few more commits.

I'll wait for you response about the label types before attempting to fix that.

I also notice that on nightly rust, `cargo doc` gives warnings like:
```
warning: field `span` is never read                                                                                                                                                                                                           
   --> plugin/src/arch/x64/ast.rs:337:9                                                                                                                                                                                                       
    |                                                                                                                                                                                                                                         
336 |     Indirect {                                                                                                                                                                                                                          
    |     -------- field in this variant                                                                                                                                                                                                      
337 |         span: Span,                                                                                                                                                                                                                     
    |         ^^^^                                                                                                                                                                                                                            
    |
    = note: `#[warn(dead_code)]` on by default

warning: field `span` is never read
   --> plugin/src/arch/x64/ast.rs:345:9
    |
344 |     Direct {
    |     ------ field in this variant
345 |         span: Span,
    |         ^^^^
...
```

But removing them also isn't correct. I'm not sure about this...

**Comment by CensoredUsername** [comment]
I'll likely look over this in the weekend. in the meantime, you should be able to generate the docs by running ./build_docs.sh

**Comment by vext01** [comment]
Let me know and I can push changes to fix whatever I've misunderstood.

File: doc/langref_common.md (hunk 1 above)
Comment (CensoredUsername) [reply doc/langref_common.md:120]: Sorry for the delay, life has a way.

Anyway, `extern` doesn't need a lookup. if you assemble `jmp extern 0x1234` it'll literally insert a jump to 0x1234. This is mostly relevant to x86, as the default jump in x86 can access the entire address space, always. However, that jump is relative, so we need to know about it for when the buffer gets moved around.

**Comment by CensoredUsername** [comment]
Poke ;) could you fix the errors in the label documentation I mentioned?

**Comment by vext01** [comment]
My apologies. I missed your message. I'll try to get to this soon.

File: doc/langref_common.md (hunk 1 above)
Comment (vext01) [reply doc/langref_common.md]: I think https://github.com/CensoredUsername/dynasm-rs/pull/92/commits/a544deae8d946f5ae7366c238ca9f8e3df26ba2c addresses the above comments.

**Comment by vext01** [comment]
@CensoredUsername were those changes any good?

**Comment by CensoredUsername** [comment]
Heya, sorry for the late reply. I want to merge this somewhere next week, it looks good :+1:

**Comment by CensoredUsername** [comment]
I did some cleanup, and it's all merged into dev now. thanks! And apologies for the delay.

**Pull request closed without merging**